### PR TITLE
feat(event serialization): add `birthRound`, `EventDescriptor` and multiple other parents to serialized event 

### LINKED ISSUE #8976: Add AddressBookRound to Serialized Event
Requires HIP
Needed to lookup the addressbook used during event creation to validate the event.

### PR DESCRIPTION
**Pull request by edward-swirldslabs**
**Description**:
This PR adds the round created to the event serialization and represents the selfParent and collection of otherParents as `EventDescriptor`s, as described in #5224.  The class organization is not completely restructured the way 5224 indicates, and this PR does not close 5224.   

Major Changes: 
* `EventDescriptor` contains the hash, creator node id, generation, and address book round of the event it describes. 
* `BaseEventHashedData` the selfParent and list of otherParents are represented as EventDescriptors. 
* `BaseEventUnhashedData` is replaced by serializing signature. 
* The prior serialization can be reserialized for the purpose of signature validation. 

The old serialization scheme for events is preserved to support replay of the events in the preconsensus event stream (PCES).  After the new serialization is deployed to mainnet, we can begin to remove support for the old serialization scheme and continue to refactor the code with the suggestions in 5224. 

**Related issue(s)**:
Fixes #8976 

**Checklist**

- [ ] HIP Created
- [X] Manually checked to support prior version in PCES Replay

### CODE REVIEW

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java [hunk 1]
```diff
@@ -31,16 +31,36 @@
...
+         */
+        public static final int ADDRESS_BOOK_ROUND = 2;
+    }
+
+    /**
+     * The version of the software discovered during deserialization that needs to be preserved in serialization.
+     * <p>
+     * DEPRECATED: Remove after 0.45 is delivered to mainnet.
+     */
+    private int deserializedVersion = ClassVersion.ADDRESS_BOOK_ROUND;
```
Comment (artemananiev) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:62]: Why is it needed? Usually the idea is to serialize according to the "current" (latest) version, while during deserialization support a few previous versions with default values for the newly added fields. I mean, if an object is deserialized from an old version, it will have some fields null / default, but when it needs to be serialized again, there is no need to serialize it in the old format

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 2]
```diff
@@ -184,13 +161,20 @@ public void serialize(
...
-        // Changing the event format may require a HIP.  The old format is preserved for now.
-        out.writeLong(creatorId.id());
-        out.writeLong(selfParentGen);
-        out.writeLong(otherParentGen);
-        out.writeSerializable(selfParentHash, false);
-        out.writeSerializable(otherParentHash, false);
+        if (deserializedVersion < ClassVersion.ADDRESS_BOOK_ROUND) {
+            // FUTURE WORK: The creatorId should be a selfSerializable NodeId at some point.
+            // Changing the event format may require a HIP.  The old format is preserved for now.
+            out.writeLong(creatorId.id());
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:167]: can we make this change now? so that `NodeId` is serialized?

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 3]
```diff
@@ -319,7 +309,10 @@ public long getClassId() {
 
     @Override
     public int getVersion() {
-        return ClassVersion.SOFTWARE_VERSION;
+        if (deserializedVersion < ClassVersion.ADDRESS_BOOK_ROUND) {
+            return deserializedVersion;
+        }
+        return ClassVersion.ADDRESS_BOOK_ROUND;
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:315]: why not just return `deserializedVersion`?

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 4]
```diff
@@ -342,28 +335,125 @@ public NodeId getCreatorId() {
...
+
+    /**
+     * Get the self parent generation
+     *
+     * @return the self parent generation
+     */
     public long getSelfParentGen() {
-        return selfParentGen;
+        if (selfParent == null) {
+            return -1;
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:372]: use `EventConstants.GENERATION_UNDEFINED`

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 5]
```diff
@@ -342,28 +335,125 @@ public NodeId getCreatorId() {
...
+    /**
+     * Get the maximum generation of the other parents.
+     *
+     * @return the maximum generation of the other parents
+     * @deprecated this method should be replaced since there can be multiple other parents.
+     */
     public long getOtherParentGen() {
-        return otherParentGen;
+        if (otherParents == null || otherParents.isEmpty()) {
+            return -1;
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:385]: use `EventConstants.GENERATION_UNDEFINED`

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 6]
```diff
@@ -393,4 +483,13 @@ public long getGeneration() {
     public static long calculateGeneration(final long selfParentGeneration, final long otherParentGeneration) {
         return 1 + Math.max(selfParentGeneration, otherParentGeneration);
     }
+
+    /**
+     * Get an event descriptor for this event.
+     *
+     * @return an event descriptor for this event
+     */
+    public EventDescriptor getEventDescriptor() {
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:492]: probably should be called `createEventDescriptor` or something, since its not a getter

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java [hunk 7]
```diff
@@ -72,22 +92,33 @@ public BaseEventUnhashedData(@Nullable final NodeId otherId, @NonNull final byte
...
+            creatorSeq = in.readLong();
+            otherId = NodeId.deserializeLong(in, true);
+            otherSeq = in.readLong();
+            signature = in.readByteArray(MAX_SIG_LENGTH);
+        } else {
+            signature = in.readByteArray(MAX_SIG_LENGTH);
+            // initialize unused fields (can be removed when the unused fields are removed)
+            otherId = null;
+            creatorSeq = SEQUENCE_UNUSED;
+            otherSeq = SEQUENCE_UNUSED;
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:120]: I think we can remove them now

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java [hunk 8]
```diff
@@ -36,8 +37,21 @@
  */
 public class GossipEvent implements BaseEvent, ChatterEvent {
     private static final long CLASS_ID = 0xfe16b46795bfb8dcL;
-    private BaseEventHashedData hashedData;
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:39]: why don't we just make the signature part of this class and get rid of `BaseEventUnhashedData` & `BaseEventImpl `?

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/CriticalQuorumTest.java [hunk 9]
```diff
@@ -78,8 +81,17 @@ public String toString() {
...
+        final EventDescriptor otherDescriptor = new EventDescriptor(new Hash(), new NodeId(0), 0, -1);
         final BaseEventHashedData baseEventHashedData = new BaseEventHashedData(
-                new BasicSoftwareVersion(1), creatorId, 0, 0, (byte[]) null, null, Instant.now(), null);
+                new BasicSoftwareVersion(1),
+                creatorId,
+                selfDescriptor,
+                Collections.singletonList(otherDescriptor),
+                -1,
+                Instant.now(),
+                null);
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/CriticalQuorumTest.java:94]: you can use `GossipEventBuilder` for this

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/EventImplTests.java [hunk 10]
```diff
@@ -238,14 +240,17 @@ void testConsensusReached() {
...
                 new BaseEventHashedData(
                         new BasicSoftwareVersion(1),
                         new NodeId(0L),
-                        0L,
-                        0L,
-                        CryptographyHolder.get().getNullHash(),
-                        CryptographyHolder.get().getNullHash(),
+                        selfDescriptor,
+                        Collections.singletonList(otherDescriptor),
+                        -1,
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/EventImplTests.java:253]: `GossipEventBuilder`

Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/EventImplTests.java:253]: in other places as well

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/signed/StartupStateUtilsTests.java [hunk 11]
```diff
@@ -128,6 +135,30 @@ private SignedState writeState(
             final boolean corrupted)
             throws IOException {
 
+        final EventImpl[] events = new EventImpl[random.nextInt(1, 10)];
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/signed/StartupStateUtilsTests.java:138]: why was this added?

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/GossipEventTest.java [hunk 12]
```diff
@@ -51,6 +56,21 @@ void test() throws IOException, ConstructableRegistryException {
         assertEquals(gossipEvent, copy, "deserialized version should be the same");
     }
 
+    @Test
+    @DisplayName("Deserialize prior version of event")
+    void deserializePriorVersion() throws IOException, ConstructableRegistryException {
```
Comment (lpetrovic05) [inline platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/GossipEventTest.java:61]: should we also check the signature if its valid?

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/resources/eventFiles/eventSerializationV45/sampleEvent.evts
Comment (lpetrovic05) [inline platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/resources/eventFiles/eventSerializationV45/sampleEvent.evts:1]: this file appears twice in different locations...

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java (hunk 1 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:62]: Supporting the prior serialization scheme is needed due to these factors: 
*  NEW: At startup, events are replayed from the saved pre-consensus event stream which are in the older serialization scheme
*  These replayed events must go through validation, including signature validation
*  This validation requires hashing the events and obtaining the exact same signature
*  Hashing uses the serialized format, so the events must be serializable in the older format to derive the saved signature.

Not doing the above leads to ISS. 

This is only relevant when we make a change to the serialization scheme, that we have to support the older serialization scheme until all events are serialized in the new serialization scheme. 

The code is configured to only serialize in the older serialization scheme if the data was deserialized from the older serialization scheme.   All new events will be serialized in the new serialization scheme.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java (hunk 3 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:315]: fair enough, I will make that change.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java (hunk 2 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:167]: I should have deleted the comment which is in the section of the code for the older serialization scheme.   The body of the else does serialize the node id as a SelfSerializable.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java (hunk 4 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:372]: Or we just use common and move the event related implementations and constants to common.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java (hunk 5 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:385]: circular dependency

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java (hunk 4 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:372]: I would say move the constants to common

Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:367]: will do

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java (hunk 6 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java]: changed

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventImpl.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+import com.swirlds.common.io.streams.SerializableDataOutputStream;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.io.IOException;
+import java.util.Objects;
+
+/**
+ * This class implements a BaseEvent which pairs hashed data with the signature used to create the hash of the data.
+ */
+public class BaseEventImpl implements BaseEvent, OptionalSelfSerializable<EventSerializationOptions> {
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventImpl.java]: removed

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java (hunk 7 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:120]: removed

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java (hunk 8 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:39]: The older serialization scheme has a version number of BaseEventUnhashedData present.  If I can remove it use in the new serialization scheme.  It's easier to keep the class until it is completely not used and then remove the class once the serialization scheme is not related to it.

Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:39]: It's going to be a lot of work to remove the use of the class, touches a lot of code that I haven't touched yet.  If you want me to do it, I will, but I've made the new serialization scheme for GossipEvent without it.

Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:39]: I think its ok to keep it for migration, but then depreciate it and remove it in a subsequent version

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 14]
```diff
@@ -342,28 +330,125 @@ public NodeId getCreatorId() {
...
+        return selfParent.getGeneration();
     }
 
+    /**
+     * Get the maximum generation of the other parents.
+     *
+     * @return the maximum generation of the other parents
+     * @deprecated this method should be replaced since there can be multiple other parents.
+     */
     public long getOtherParentGen() {
```
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:378]: It's not immediately obvious to me if we will want to have this method return the maximum other parent generation. We may not even keep this method around when we actually support multiple other parents. But I'm ok punting this design decision until later.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 15]
```diff
@@ -342,28 +330,125 @@ public NodeId getCreatorId() {
...
+        return selfParent.getHash();
     }
 
+    /**
+     * Get the hash of the other parent with the maximum generation.
+     *
+     * @return the hash of the other parent with the maximum generation
+     * @deprecated
+     * */
     public Hash getOtherParentHash() {
```
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:409]: Once again, not sure we will keep this method when we start using multiple other parents. But not harmful for now.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java [hunk 16]
```diff
@@ -145,7 +176,31 @@ public NodeId getOtherId() {
...
+     * Get the signature
+     *
+     * @return the signature
+     */
     public byte[] getSignature() {
         return signature;
     }
+
+    /**
+     * Synchronize the creator data between the hashed event's other parent data and the creatorId.  T his method is to
```
Comment (cody-littley) [inline platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:189]:
```suggestion
     * Synchronize the creator data between the hashed event's other parent data and the creatorId.  This method is to
```

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/CriticalQuorumTest.java (hunk 9 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/CriticalQuorumTest.java:94]: I'd recommend you don't spend too much time on this test. Critical quorum is disabled in `0.44`, and I intend to delete it in `0.45`.

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/signed/StartupStateUtilsTests.java (hunk 11 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/state/signed/StartupStateUtilsTests.java:138]: removed, good catch.  It must have been a stray copy-paste.

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/GossipEventTest.java (hunk 12 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/GossipEventTest.java:61]: The most important thing is that the bytes on the wire are preserved.  I no longer know which public key produced the event for the purpose of validation.  I've extended the unit test to check the bytes serialized are identical to the file contents.

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/resources/eventFiles/eventSerializationV45/sampleEvent.evts
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/resources/eventFiles/eventSerializationV45/sampleEvent.evts:1]: I am renaming the file for clarity, they contain different contents.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 17]
```diff
@@ -221,18 +203,33 @@ public void deserialize(
...
-        // FUTURE WORK: The creatorId should be a selfSerializable NodeId at some point.
-        // Changing the event format may require a HIP.  The old format is preserved for now.
-        creatorId = NodeId.deserializeLong(in, false);
-        selfParentGen = in.readLong();
-        otherParentGen = in.readLong();
-        selfParentHash = in.readSerializable(false, Hash::new);
-        otherParentHash = in.readSerializable(false, Hash::new);
+        if (version < ClassVersion.ADDRESS_BOOK_ROUND) {
+            // FUTURE WORK: The creatorId should be a selfSerializable NodeId at some point.
+            // Changing the event format may require a HIP.  The old format is preserved for now.
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:218]: I am informed

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 18]
```diff
@@ -342,28 +330,125 @@ public NodeId getCreatorId() {
...
+    public EventDescriptor getSelfParent() {
+        return selfParent;
+    }
+
+    /**
+     * Get the event descriptors for the other parents.
+     *
+     * @return the event descriptors for the other parents
+     */
+    public List<EventDescriptor> getOtherParents() {
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:363]: Added return value annotations to all methods.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 19]
```diff
@@ -221,18 +203,33 @@ public void deserialize(
...
+            otherParents = otherParentHash == null
+                    ? null
+                    : Collections.singletonList(new EventDescriptor(otherParentHash, otherParentGen, -1));
+            addressBookRound = -1;
+        } else {
+            creatorId = in.readSerializable(false, NodeId::new);
+            selfParent = in.readSerializable(false, EventDescriptor::new);
+            otherParents = in.readSerializableList(AddressBook.MAX_ADDRESSES, false, EventDescriptor::new);
+            addressBookRound = in.readLong();
+        }
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:239]: Coordinated.  I'll absorb the conflicting changes when his PR merges into develop.   Adding throw of IOException when non-nullable fields are deserialized as null.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java (hunk 14 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:385]: Agreed.  I will throw an unsupported operation exception if there are multiple other parents so that we are forced to rewrite this method when we actually start using multiple other parents.

File: platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/EventImplTests.java (hunk 10 above)
Comment (lpetrovic05) [reply platform-sdk/swirlds-platform-core/src/test/java/com/swirlds/platform/EventImplTests.java:253]: thats a nightmare.... let's not delay this PR because of this

**Review by lpetrovic05 — APPROVED** [review]

**Review by lpetrovic05 — DISMISSED** [review]

**Review by cody-littley — DISMISSED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java [hunk 20]
```diff
@@ -68,20 +84,31 @@ public GossipEvent(final BaseEventHashedData hashedData, final BaseEventUnhashed
...
+        serializedVersion = version;
+        if (version < ClassVersion.ROSTER_ROUND) {
+            hashedData = in.readSerializable(false, BaseEventHashedData::new);
+            unhashedData = in.readSerializable(false, BaseEventUnhashedData::new);
+        } else {
+            hashedData = in.readSerializable(false, BaseEventHashedData::new);
+            final byte[] signature = in.readByteArray(MAX_SIG_LENGTH);
+            unhashedData = new BaseEventUnhashedData(null, signature);
         }
+        // remove update of other parent event descriptor after 0.45.0 hits mainnet.
```
Comment (alittley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:110]: Once 45 hits mainnet, would we be able to delete `BaseEventUnhashedData` entirely, and just directly use the signature?

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/tipset/TipsetEventCreatorTests.java [hunk 21]
```diff
@@ -152,9 +153,11 @@ private void validateNewEvent(
             final boolean slowNode) {
 
         final EventImpl selfParent = events.get(newEvent.getHashedData().getSelfParentHash());
-        final long selfParentGeneration = selfParent == null ? -1 : selfParent.getGeneration();
+        final long selfParentGeneration =
+                selfParent == null ? -EventConstants.GENERATION_UNDEFINED : selfParent.getGeneration();
```
Comment (alittley) [inline platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/tipset/TipsetEventCreatorTests.java:157]: Did you intend to but a `-` sign in front of the constant?

**Review by alittley — APPROVED** [review]

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/tipset/TipsetEventCreatorTests.java (hunk 21 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/tipset/TipsetEventCreatorTests.java:157]: nope :-)

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java (hunk 20 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/event/GossipEvent.java:110]: there is more refactor to perform to get rid of `BaseEventUnhashedData` so that nothing is referencing it.   But after this pr hits mainnet, there should no longer be instances of it in the serialization scheme and we're free to proceed with excising the tumor from the code.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 22]
```diff
@@ -67,7 +70,24 @@ private static class ClassVersion {
          * In this version, the software version of the node that created this event is included in the event.
          */
         public static final int SOFTWARE_VERSION = 3;
+
+        /**
+         * Event descriptors replace the hashes and generation of the parents in the event.
+         * Multiple otherParents are supported.
+         * rosterRound is added for lookup of the effective roster at the time of event creation.
+         *
+         * @since 0.44.0
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:79]: Correct.  thanks!

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 23]
```diff
@@ -117,63 +133,24 @@ public BaseEventHashedData() {}
     public BaseEventHashedData(
             @NonNull SoftwareVersion softwareVersion,
             @NonNull final NodeId creatorId,
-            final long selfParentGen,
-            final long otherParentGen,
-            @Nullable final Hash selfParentHash,
-            @Nullable final Hash otherParentHash,
+            @Nullable final EventDescriptor selfParent,
+            @Nullable final List<EventDescriptor> otherParents,
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:137]: I'll implement this change.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 24]
```diff
@@ -117,63 +133,24 @@ public BaseEventHashedData() {}
...
             @NonNull final Instant timeCreated,
             @Nullable final ConsensusTransactionImpl[] transactions) {
         this.softwareVersion = Objects.requireNonNull(softwareVersion, "The softwareVersion must not be null");
         this.creatorId = Objects.requireNonNull(creatorId, "The creatorId must not be null");
-        this.selfParentGen = selfParentGen;
-        this.otherParentGen = otherParentGen;
-        this.selfParentHash = selfParentHash;
-        this.otherParentHash = otherParentHash;
+        this.selfParent = selfParent;
+        if (otherParents != null && otherParents.size() == 1 && otherParents.get(0) == null) {
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:144]: will do.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java [hunk 25]
```diff
@@ -184,13 +161,19 @@ public void serialize(
...
-        // FUTURE WORK: The creatorId should be a selfSerializable NodeId at some point.
-        // Changing the event format may require a HIP.  The old format is preserved for now.
-        out.writeLong(creatorId.id());
-        out.writeLong(selfParentGen);
-        out.writeLong(otherParentGen);
-        out.writeSerializable(selfParentHash, false);
-        out.writeSerializable(otherParentHash, false);
+        if (serializedVersion < ClassVersion.ROSTER_ROUND) {
+            out.writeLong(creatorId.id());
+            out.writeLong(selfParent != null ? selfParent.getGeneration() : EventConstants.ROSTER_ROUND_UNDEFINED);
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventHashedData.java:166]: good catch

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java [hunk 26]
```diff
@@ -145,7 +164,31 @@ public NodeId getOtherId() {
...
+     *
+     * @return the signature
+     */
     public byte[] getSignature() {
         return signature;
     }
+
+    /**
+     * Synchronize the creator data between the hashed event's other parent data and the creatorId.  This method is to
+     * support backwards compatibility with the previous event serializaiton format.
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:178]: good catch, thanks!

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java [hunk 27]
```diff
@@ -102,26 +123,19 @@ public boolean equals(final Object o) {
...
     }
 
     @Override
     public String toString() {
         final int signatureLength = signature == null ? 0 : signature.length;
-        return "BaseEventUnhashedData{" + "creatorSeq="
-                + creatorSeq + ", otherId="
-                + otherId + ", otherSeq="
-                + otherSeq + ", signature="
+        return "BaseEventUnhashedData{" + ", otherId="
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:137]: replaced with usage of ToStringBuilder.  Also replaced EventDescriptor.toString implementation with it.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java [hunk 28]
```diff
@@ -145,7 +164,31 @@ public NodeId getOtherId() {
...
+    /**
+     * Synchronize the creator data between the hashed event's other parent data and the creatorId.  This method is to
+     * support backwards compatibility with the previous event serializaiton format.
+     *
+     * @param event the event to update
+     * @deprecated This method should be deleted when we are no longer supporting the previous event serialization
+     * format.
+     */
+    public void updateOtherParentEventDescriptor(@NonNull final BaseEventHashedData event) {
+        if (otherId != null && event.hasOtherParent()) {
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/BaseEventUnhashedData.java:185]: hrm..   you're right.  Fixing the logic and re-testing to see if anything was dependent on it being wrong.

File: platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/EventDescriptor.java [hunk 29]
```diff
@@ -177,7 +238,8 @@ public int hashCode() {
     @Override
     public String toString() {
         return "(creator: " + creator + ", generation: "
```
[the remark this answers is not in the document]
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-common/src/main/java/com/swirlds/common/system/events/EventDescriptor.java:240]: it is now.

File: platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/tipset/TipsetEventCreatorTests.java (hunk 21 above)
Comment (edward-swirldslabs) [reply platform-sdk/swirlds-unit-tests/core/swirlds-platform-test/src/test/java/com/swirlds/platform/test/event/tipset/TipsetEventCreatorTests.java:157]: find replace failure.

**Review by alittley — APPROVED** [review]

**Review by alittley — APPROVED** [review]

**Review by alittley — APPROVED** [review]

**Review by cody-littley — APPROVED** [review]

**Review by lpetrovic05 — DISMISSED** [review]

**Review by cody-littley — APPROVED** [review]

**Review by alittley — APPROVED** [review]
